src/storage/db/rocksdb.rs

```diff
@@ -0,0 +1,541 @@
+use crate::storage::store::increment_vec_u8;
+use prost::DecodeError;
+use rocksdb::{Options, TransactionDB};
+use std::collections::HashMap;
+use std::fs::{self};
+use std::path::Path;
+use std::sync::{RwLock, RwLockReadGuard};
+use thiserror::Error;
+use tracing::info;
+use walkdir::WalkDir;
+
+#[derive(Error, Debug)]
+pub enum RocksdbError {
+    #[error(transparent)]
+    InternalError(#[from] rocksdb::Error),
+
+    #[error(transparent)]
+    DecodeError(#[from] DecodeError),
+
+    #[error("DB is not open")]
+    DbNotOpen,
+}
+
+/** Hold a transaction. List of key/value pairs that will be committed together */
+pub struct RocksDbTransactionBatch {
+    pub batch: HashMap<Vec<u8>, Option<Vec<u8>>>,
+}
+
+impl RocksDbTransactionBatch {
+    pub fn new() -> RocksDbTransactionBatch {
+        RocksDbTransactionBatch {
+            batch: HashMap::new(),
+        }
+    }
+
+    pub fn put(&mut self, key: Vec<u8>, value: Vec<u8>) {
+        self.batch.insert(key, Some(value));
+    }
+
+    pub fn delete(&mut self, key: Vec<u8>) {
+        self.batch.insert(key, None);
+    }
+
+    pub fn merge(&mut self, other: RocksDbTransactionBatch) {
+        for (key, value) in other.batch {
+            self.batch.insert(key, value);
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.batch.len()
+    }
+}
+
+pub struct IteratorOptions {
+    pub opts: rocksdb::ReadOptions,
+    pub reverse: bool,
+}
+
+#[derive(Default)]
+pub struct RocksDB {
+    pub db: RwLock<Option<rocksdb::TransactionDB>>,
+    pub path: String,
+}
+
+#[derive(Debug, Default)]
+pub struct PageOptions {
+    pub page_size: Option<usize>,
+    pub page_token: Option<Vec<u8>>,
+    pub reverse: bool,
+}
+
+impl std::fmt::Debug for RocksDB {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RocksDB").field("path", &self.path).finish()
+    }
+}
+
+impl RocksDB {
```

Comment:
Is this copied as is from the hub? Any notable changes?

Reply:
Added a note on material changes in the description.